src/Evolution/Systems/ForceFree/MaskNeutronStarInterior.hpp

```diff
@@ -0,0 +1,89 @@
+#include "PointwiseFunctions/AnalyticData/ForceFree/Factory.hpp"
+#include "PointwiseFunctions/AnalyticSolutions/ForceFree/Factory.hpp"
+#include "PointwiseFunctions/InitialDataUtilities/InitialData.hpp"
+#include "PointwiseFunctions/InitialDataUtilities/Tags/InitialData.hpp"
+#include "Utilities/CallWithDynamicType.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/ProtocolHelpers.hpp"
+#include "Utilities/SetNumberOfGridPoints.hpp"
+#include "Utilities/TMPL.hpp"
+
+namespace ForceFree {
+/*!
+ * \brief Assign the masking scalar variable (see Tags::NsInteriorMask) at the
+ * initialization phase in NS magnetosphere simulations.
+ *
+ * For an isolated NS magnetosphere simulations, center of the NS is located at
+ * the origin and coordinate radius of the NS is rescaled to 1. Grid points
+ * with $x^2 + y^2 + z^2 < 1.0$ are identified as interior, and are subject to
+ * correction of evolved variables in each time steps to enforce the MHD
+ * condition (see ImposeMhdConditionInsideNs)
+ *
+ * \note The location of NS interior (mask) is fixed in the Grid frame to avoid
+ * the NS surface crossing grid points during the evolution.
+ *
+ */
+template <typename Metavariables, bool UsedForFdGrid>
+struct MaskNeutronStarInterior : tt::ConformsTo<db::protocols::Mutator> {
+  using argument_tags =
+      tmpl::list<tmpl::conditional_t<
+                     UsedForFdGrid,
+                     evolution::dg::subcell::Tags::Coordinates<3, Frame::Grid>,
+                     domain::Tags::Coordinates<3, Frame::Grid>>,
+                 evolution::initial_data::Tags::InitialData>;
+
+  using return_tags = tmpl::list<tmpl::conditional_t<
+      UsedForFdGrid,
+      evolution::dg::subcell::Tags::Inactive<Tags::NsInteriorMask>,
+      Tags::NsInteriorMask>>;
+
+  static void apply(
+      const gsl::not_null<std::optional<Scalar<DataVector>>*>
+          neutron_star_interior_mask,
+      const tnsr::I<DataVector, 3, Frame::Grid>& grid_frame_coords,
+      const evolution::initial_data::InitialData& solution_or_data) {
+    using all_data_and_solutions =
+        tmpl::at<typename Metavariables::factory_creation::factory_classes,
+                 evolution::initial_data::InitialData>;
+
+    using ns_magnetosphere_id_list =
+        tmpl::append<ForceFree::AnalyticData::available_data<
+                         SimulationType::IsolatedNsMagnetosphere>,
+                     ForceFree::Solutions::available_solutions<
+                         SimulationType::IsolatedNsMagnetosphere>>;
+
+    call_with_dynamic_type<void, all_data_and_solutions>(
+        &solution_or_data, [&neutron_star_interior_mask,
+                            &grid_frame_coords](const auto* initial_data_ptr) {
+          using InitialData = std::decay_t<decltype(*initial_data_ptr)>;
+
+          if constexpr (tmpl::list_contains_v<ns_magnetosphere_id_list,
+                                              InitialData>) {
```

Comment:
Do the call unconditionally.  I think if you use `Tags::AnalyticSolutionOrData` in `argument_tags` you can even avoid the `call_with_dynamic_type`.  The function in `src/Evolution/Initialization/InitialData.hpp` might be useful, but I'm not sure on that one.

Reply:
1. `Forcefree` system is using runtime initial data based on the factory infrastructure, which was why I used `call_with_dynamic_type`.
2. If run unconditionally, wouldn't this raise "tag is not retrievable" error for other types of initial data? e.g. simple 1D wave tests.

Reply:
1. Yes, you're right.  I was looking at code that hasn't been converted to run-time yet.
2. You'll have to either implement it for the others, or make it it's own function instead of a `variables` specialization and check if that function is callable.

Reply:
Okay, I'll change this to something other than the `variables` specialization.

Reply:
Use `CREATE_IS_CALLABLE` instead of hardcoding the solutions that need masks.